[Xcode] Simplify sanitizer builds

ddkilzer opened the pull request:
#### 5ee7a204c4e90b89cb781ade596d36a816043c29
<pre>
[Xcode] Simplify sanitizer builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=244540">https://bugs.webkit.org/show_bug.cgi?id=244540</a>
&lt;rdar://98015260&gt;

Reviewed by Elliott Williams.

This change simplifies sanitizer builds of WebKit using Xcode.
Instead of specifying `-xcconfig path/to/*san.xcconfig` on the
command-line, the ENABLE_*_SANITIZER=YES Xcode variables now
activate WebKit-specific settings.

All sanitizer settings are now in a single file named
Configurations/Sanitizers.xcconfig, which is included in every
project via Configurations/CommonBase.xcconfig.

Changes to Base.xcconfig files:
- Rename GCC_OPTIMIZATION_LEVEL variables to add WK_DEFAULT_
  prefix to support overriding in CommonBase.xcconfig.
- OTHER_CFLAGS, OTHER_CPLUSPLUS_FLAGS and OTHER_LDFAGS variables
  are defined in CommonBase.xcconfig, and per-project
  Base.xccconfig files now pick up those values through use of
  $(inherited).  When $(inherited) is not used, the
  corresponding WK_COMMON_ variable is used to include these
  values in various targets.

Other changes too numerous to list individually below:
- Rename all sanitizer variables to add WK_ prefixes.
- Rename ENABLE_ANY_SANITIZER to WK_ANY_SANITIZER_ENABLED in
  WK_NO_STATIC_INITIALIZERS variables.
- Rename WK_ASAN_DISALLOWED to WK_SANITIZER_DISALLOWED in
  yasm.xcconfig and ImageDiff.xcconfig since it disables all
  sanitizers.
- Remove ASAN_OTHER_LDFLAGS variable when the value of
  WK_SANITIZER_OTHER_LDFLAGS is inherited.
- Replace ASAN_OTHER_LDFLAGS with WK_COMMON_OTHER_LDFAGS when
  OTHER_LDFLAGS does not use $(inherited).
- Move WK_LIBCPP_ASSERTIONS_CFLAGS to be added to
  OTHER_CPLUSPLUSFLAGS in CommonBase.xcconfig.  This enables the
  macro in more projects, and removes it from OTHER_CFLAGS in
  some projects (since it only affects C++ code).

* Configurations/CommonBase.xcconfig:
- Include Sanitizers.xcconfig.
- GCC_OPTIMIZATION_LEVEL is now set (or overridden) here instead
  of in each project&apos;s Base.xcconfig.
- Initial values of OTHER_CFLAGS, OTHER_CPLUSPLUS_FLAGS and
  OTHER_LDFAGS variables are now set here.
- Move WK_COMMON_WARNING_CFLAGS and add FIXME comment.
* Configurations/SDKVariant.xcconfig:
- Add WK_ADD_ macros used by Sanitizers.xcconfig.
* Configurations/Sanitizers.xcconfig: Add.
- Combines all of the Tools/sanitizer xcconfig files into one.
- Add -fsanitize=enum,return for UBSan since some versions of
  Xcode disable these checks by default.
- Add a FIXME for disabling sanitizers on host-side tools.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
- Switch OTHER_LDFLAGS to use $(inherited).
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/ThirdParty/libwebrtc/Configurations/yasm.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/libavif.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/dav1d.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/Configurations/webpushd.xcconfig:
* Source/WebKit/Configurations/webpushtool.xcconfig:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/DumpRenderTree/mac/Configurations/BaseTarget.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeApp.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
- Delete OTHER_LDFLAGS since it was not overriding anything from
  its Base.xcconfig.
* Tools/ImageDiff/cg/Configurations/ImageDiff.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/WebGPUPlayground/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

* Tools/Scripts/webkitdirs.pm:
(XcodeOptions):
- Simplify options for enabling sanitizers or overriding the
  compiler optimization level.
* Tools/Scripts/webkitpy/style/checker.py:
(_all_categories):
- Add BaseXcconfigChecker categories.
(FileType):
- Replace unused SDKVARIANT with BASE_XCCONFIG.
(CheckerDispatcher._file_type):
- Return FileType for Base.xcconfig files.
(CheckerDispatcher._create_checker):
- Set checker for Base.xcconfig files.
* Tools/Scripts/webkitpy/style/checker_unittest.py:
(CheckerDispatcherDispatchTest.assert_checker_basexcconfig):
(CheckerDispatcherDispatchTest.test_basexcconfig_paths):
- Add test for BaseXcconfigChecker.
* Tools/Scripts/webkitpy/style/checkers/basexcconfig.py: Add.
(BaseXcconfigChecker):
(BaseXcconfigChecker.__init__):
(BaseXcconfigChecker.read_common_base_xcconfig_variables):
(BaseXcconfigChecker.check):
- Implement BaseXcconfigChecker.
- The checker is dynamic in that it reads CommonBase.xcconfig to
  determine which variables require checking in Base.xcconfig
  project files.
* Tools/Scripts/webkitpy/style/checkers/basexcconfig_unittest.py: Add.
(BaseXcconfigCheckerTest):
(BaseXcconfigCheckerTest.assert_no_error):
(BaseXcconfigCheckerTest.assert_no_error.handle_style_error):
(BaseXcconfigCheckerTest.assert_error):
(BaseXcconfigCheckerTest.assert_error.handle_style_error):
(BaseXcconfigCheckerTest.mock_handle_style_error):
(BaseXcconfigCheckerTest.test_init):
(BaseXcconfigCheckerTest.test_gcc_optimization_level):
(BaseXcconfigCheckerTest.test_wk_default_gcc_optimization_level):
(BaseXcconfigCheckerTest.test_other_cflags):
(BaseXcconfigCheckerTest.test_other_cplusplusflags):
(BaseXcconfigCheckerTest.test_other_ldflags):
- Add unit tests for BaseXcconfigChecker.

* Tools/sanitizer/asan+ubsan.xcconfig: Remove.
* Tools/sanitizer/asan.xcconfig: Remove.
* Tools/sanitizer/sanitizer.xcconfig: Remove.
* Tools/sanitizer/tsan.xcconfig: Remove.
* Tools/sanitizer/ubsan.xcconfig: Remove.

Canonical link: <a href="https://commits.webkit.org/260654@main">https://commits.webkit.org/260654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df92739a6d3ea57b490bd03feccb1ccd8cd8fd12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108987 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/520 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19519 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9363 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101221 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114744 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/112490 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10850 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/8959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11596 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/50459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/106488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13194 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/106488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4017 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -0,0 +1,78 @@
+
+OTHER_CPLUSPLUSFLAGS = $(inherited) $(WK_ADDRESS_SANITIZER_OTHER_CPLUSPLUSFLAGS_$(ENABLE_ADDRESS_SANITIZER));
+
+OTHER_LDFLAGS = $(inherited) $(WK_ANY_SANITIZER_LDFLAGS_$(WK_ANY_SANITIZER_ENABLED));
```
These settings will be silently overridden by any project's Base.xcconfig that also defines them. For example, OTHER_CPLUSPLUSFLAGS will use this definition in JavaScriptCore: https://github.com/WebKit/WebKit/blob/91f7b27a2c241b8753006ec66762aacbd2c66089/Source/JavaScriptCore/Configurations/Base.xcconfig#L176

When run with ENABLE_ADDRESS_SANITIZER=YES, this is what happens:
```
> xcodebuild -project Source/JavaScriptCore/JavaScriptCore.xcodeproj -showBuildSettings ENABLE_ADDRESS_SANITIZER=YES | grep -E "(WK_.*_SANITIZER|OTHER_.*
FLAGS)"                                                                                                                                                                               
    OTHER_CFLAGS =  -DRELEASE_WITHOUT_OPTIMIZATIONS -fno-omit-frame-pointer -g  
    OTHER_CPLUSPLUSFLAGS =  -DRELEASE_WITHOUT_OPTIMIZATIONS -fno-omit-frame-pointer -g   -isystem <SDKROOT>/System/Library/Frameworks/System.framework/PrivateHeaders
    OTHER_LDFLAGS =  -Wl,-rpath,@executable_path/Frameworks
    WK_ADDRESS_SANITIZER_OTHER_CPLUSPLUSFLAGS_YES = -U_LIBCPP_HAS_NO_ASAN
    WK_ANY_SANITIZER_CFLAGS_YES = -DRELEASE_WITHOUT_OPTIMIZATIONS -fno-omit-frame-pointer -g
    WK_ANY_SANITIZER_ENABLED = YES
    WK_ANY_SANITIZER_LDFLAGS_YES = -Wl,-rpath,@executable_path/Frameworks
    WK_UNDEFINED_BEHAVIOR_SANITIZER_OTHER_CFLAGS_YES = -fno-delete-null-pointer-checks -fno-optimize-sibling-calls -fno-sanitize=vptr -fsanitize=enum,return
```

* JSC doesn't define `OTHER_CFLAGS`, so the definition from Sanitizers.xcconfig is used.
* At the SDK level, `OTHER_CPLUSPLUSFLAGS` includes `$(OTHER_CFLAGS),` so the `$(inherited)` clause on JSC's `OTHER_CPLUSPLUSFLAGS` declaration pulls it in.
  * But it doesn't inherit the (now overridden) definition from Sanitizers.xcconfig.

---

Fundamentally, the problem here is that `$(inherited)` inherits values from different _layers_ of configurations, not from previous declarations.

For our purposes, if we're going to define sanitizer settings in the project layer, I think we'll need to either:

* collect all the sanitizer flags we'll need into one variable (`WK_SANITIZER_CFLAGS` et al.), then use that variable to set `OTHER_CFLAGS` et al. in each project's Base.xcconfig
* set `OTHER_CFLAGS` et al. here, and introduce variables like `WK_OTHER_CFLAGS` which projects can set to add in additional flags
  * If we do this, we'd probably want a style checker rule to prevent Base.xcconfigs from accidentally overriding these.

What do you think?

ddkilzer replied:
This is awesome.  I never realized that `$(inherited)` only worked between different layers.  That makes so much sense now.

My current plan is to define `OTHER_CFLAGS`, `OTHER_CPLUSPLUSFLAGS` and `OTHER_LDFLAGS` in `Configurations/CommonBase.xcconfig` (to reduce duplicate code in each `Base.xcconfig`), and add `WK_BASE_` prefix to `OTHER_*` variables in `Base.xcconfig` files.

And add a style checker rule as you suggested.

I was thinking about using `WK_BASE_` as the prefix to give a hint that it would be set in `Base.xcconfig` files if it exists.

emw-apple commented on Configurations/BaseCommon.xcconfig:
```diff
@@ -0,0 +1,41 @@
+// Define GCC_OPTIMIZATION_LEVEL_WK_OVERRIDE_NO in each project's Base.xcconfig.
+
+WK_GCC_OPTIMIZATION_LEVEL = $(WK_GCC_OPTIMIZATION_LEVEL_FORCED_$(WK_HAS_FORCE_OPTIMIZATION_LEVEL));
+WK_GCC_OPTIMIZATION_LEVEL_FORCED_YES = $(WK_FORCE_OPTIMIZATION_LEVEL);
```
What's the use case for WK_FORCE_OPTIMIZATION_LEVEL? 

I'm probably misunderstanding, but I don't see anywhere in the PR where we set this. If it's intended to be set ad-hoc by users, why can't they just override GCC_OPTIMIZATION_LEVEL on the command line?

ddkilzer replied:
It's a `Tools/Scripts/set-webkit-configuration` switch:

```
  --force-optimization-level=<level> Optimization level: O3, O2, O1, O0, Os, Ofast, Og, or none
```

Used by @TuomasWebKit and the JavaScriptCore team, IIRC.

ddkilzer replied:
And I just realized that with this change, we're no longer specifying an `xcconfig` file on the command-line, so we can simplify this code in `Tools/Scripts/webkitdirs.pm`:

```
    if ($forceOptimizationLevel) {
        if ($asanIsEnabled || $tsanIsEnabled || $ubsanIsEnabled) {
            # Command-line Xcode variable won't override that same varible set in a command-line xcconfig file.
            push @options, "WK_FORCE_OPTIMIZATION_LEVEL=$forceOptimizationLevel";
        } else {
            push @options, "GCC_OPTIMIZATION_LEVEL=$forceOptimizationLevel";
        }
    }
```

emw-apple commented on Configurations/BaseCommon.xcconfig:
```diff
@@ -0,0 +1,41 @@
+
+GCC_OPTIMIZATION_LEVEL = $(GCC_OPTIMIZATION_LEVEL_WK_OVERRIDE_$(WK_OVERRIDE_GCC_OPTIMIZATION_LEVEL));
+GCC_OPTIMIZATION_LEVEL_WK_OVERRIDE_YES = $(WK_GCC_OPTIMIZATION_LEVEL);
+// Define GCC_OPTIMIZATION_LEVEL_WK_OVERRIDE_NO in each project's Base.xcconfig.
```
I think it would be more understandable if this were named `WK_DEFAULT_GCC_OPTIMIZATION_LEVEL` or simply `WK_GCC_OPTIMIZATION_LEVEL`, though the latter would mean shuffling arround the name of the setting used below.

ddkilzer replied:
Yeah, after looking at this patch again, I think I prefer `WK_GCC_OPTIMIZATION_LEVEL` as well.  Thanks!

ddkilzer replied:
Actually, I think I'll use `WK_BASE_GCC_OPTIMIZATION_LEVEL` to match how `WK_BASE_` will be used for `OTHER_*` variables.

emw-apple reviewed (dismissed):
I really like the approach here, though we need to figure out how to avoid clobbering variables like OTHER_CFLAGS.

aproskuryakov commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -0,0 +1,85 @@
+WK_ADDRESS_SANITIZER_OTHER_CPLUSPLUSFLAGS_YES = -U_LIBCPP_HAS_NO_ASAN;
+
+// Workaround a false positive <https://bugs.webkit.org/show_bug.cgi?id=236001>.
+// Update after <rdar://87613908> is fixed.
```
It's already fixed, so the workaround can at least be made conditional, if not removed.

ddkilzer replied:
It's already conditional—see the definition of `WK_NEEDS_ASAN_USE_AFTER_SCOPE_WORKAROUND`.

I'll update the comment.

aproskuryakov reviewed (commented):
The WK_BASE_ prefix approach to overriding seems fragile, as it depends on no one ever re-adding native Xcode variables. I don't fully understand the issue being solved though, so I don't have an alternative idea.

emw-apple commented on Configurations/CommonBase.xcconfig:
```diff
@@ -23,11 +23,44 @@
+// WK_EXTRA_ prefix is used in Base.xcconfig project files for
+// multi-value variables that should be included in variable definitions
+// in CommonBase.xcconfig.
+// Example: WK_EXTRA_OTHER_{CFLAGS,CPLUSPLUSFLAGS,LDFLAGS}.
```
Hm… COMMON_OTHER and EXTRA_OTHER read as superfluous to me. What would you think about ditching the "OTHER_", e.g. `WK_COMMON_CFLAGS`, `WK_EXTRA_LDFLAGS`, etc. I think it's more fluent, and that these settings are still easy to correlate with their built-in setting names.

ddkilzer replied:
Yes, that reads better.  I'll make the change.

ddkilzer replied:
Actually, with what we found out last night, the `WK_EXTRA_` variables may not even be needed because we can just use `$(inherited)` and the same variable name in both files, so I think these will just go away.

ddkilzer replied:
After doing the next patch refactoring, I think that it's best to keep `WK_COMMON_OTHER_CFLAGS` rather than shortening it to `WK_COMMON_CFLAGS`.

There are two reasons for this:
1. I basically don't have to do any research (or think very hard) after removing `WK_COMMON_` prefix to know which variable it's providing common values for.  I only have to look at the remaining variable name instead of looking at `Configurations/CommonBase.xcconfig`.
2. It would be inconsistent with `WK_COMMON_WARNING_CFLAGS` since they are both `CFLAGS` switches.  Why keep `WARNING_` but remove `OTHER_`?

So the next patch will remove `WK_EXTRA_` altogether, but keep `WK_COMMON_` as a prefix without altering the name of the variable it's providing common values for.

emw-apple commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -0,0 +1,84 @@
+
+WK_NEEDS_ASAN_USE_AFTER_SCOPE_WORKAROUND = $(WK_NOT_$(WK_OR_$(WK_XCODE_VERSION_BEFORE_13_3_$(XCODE_VERSION_MAJOR))_$(WK_XCODE_VERSION_AFTER_13_3_$(XCODE_VERSION_MAJOR))));
+
+WK_XCODE_VERSION_BEFORE_13_3_0800 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_0900 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1000 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1100 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1200 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1300 = $(WK_XCODE_VERSION_BEFORE_13_3_1300_$(XCODE_VERSION_MINOR));
+WK_XCODE_VERSION_BEFORE_13_3_1300_1300 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1300_1310 = YES;
+WK_XCODE_VERSION_BEFORE_13_3_1300_1320 = YES;
+WK_XCODE_VERSION_AFTER_13_3_1400 = YES;
+WK_XCODE_VERSION_AFTER_13_3_1500 = YES;
+WK_XCODE_VERSION_AFTER_13_3_1600 = YES;
+WK_XCODE_VERSION_AFTER_13_3_1700 = YES;
```
Just a note, I'm putting up a PR which adds Xcode version conditionals to `WebKitTargetConditionals.xcconfig`. I see that this is moved code, so no need to try to refactor it now, but I can follow up with you when my change is available.

ddkilzer replied:
I wrote these to be very specific to the ASan workaround as it checks for Xcode versions < 13.3 or >= 14.0 since the regression happened during 13.x software updates.  (I was hoping that we could get rid of this code after we stopped building with Xcode 13.x. :)

emw-apple commented on Tools/Scripts/webkitpy/style/checkers/basexcconfig.py:
```diff
@@ -0,0 +1,62 @@
+"""Enforces rules for Base.xcconfig files."""
+
+
+class BaseXcconfigChecker(object):
```
I think this rule would be even better if it dynamically parsed setting names from `CommonBase.xcconfig`. Otherwise, I expect people will move things into CommonBase without knowing to add a clause to the rule.

At initialization, we could do something like:
```python
self.common_settings_and_extensions = {}
for line in open('Configurations/CommonBase.xcconfig'):
    # Find lines containing assignments.
    lhs, operator, rhs = map(str.strip, line.partition('='))
    # Skip non-assignment lines, comments, and WK_ variables which are allowed to be overridden.
    if operator != '=' or lhs.startswith('//') or lhs.startswith('WK_'):
        continue
    # Discard any setting condition.
    name = lhs.partition('[')[0]
    suggested_extension_name = f'WK_EXTRA_{name}' if 'WK_EXTRA_' in rhs else f'WK_DEFAULT_{name}'
    self.common_settings_and_extensions[name] = suggested_extension_name
```

ddkilzer replied:
Great idea!

With what we learned about how `$(inherited)` works with XCBuild, I think we can tweak the code above to put non-`WK_` variables that include `$(inherited)` in a different category, and then verify that `$(inherited)` also appears in the same variable(s) in `Base.xcconfig`.

ddkilzer replied:
This works so well!

emw-apple reviewed (commented):
This is a really nice improvement. Thank you so much! I have a variable name question, and a suggestion to extend the checker rule (which you are free to ignore if you think it's out of scope).

emw-apple commented on Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
```diff
@@ -80,7 +80,7 @@ ANGLE_INSTALL_PATH_PREFIX = $(ANGLE_INSTALL_PATH_PREFIX_DEPLOYMENT_$(DEPLOYMENT_
 ANGLE_INSTALL_PATH_PREFIX_DEPLOYMENT_YES_USE_ALTERNATE_YES = $(WK_ALTERNATE_FRAMEWORKS_DIR)/;
 
-OTHER_LDFLAGS = $(ASAN_OTHER_LDFLAGS) $(ANGLE_OTHER_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) $(ANGLE_OTHER_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```
Now that we're clear on how $(inherited) works in XCBuild, wouldn't it be better to do:
```suggestion
OTHER_LDFLAGS = $(inherited) $(ANGLE_OTHER_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```

ddkilzer replied:
In theory, yes.

In practice, I don't want to change behavior too much on a large patch like this because there's already a good chance something will break without changing this behavior.

If we want to make these kinds of behavior changes, they should be made outside of this patch.

Also, the fact that `$(WK_COMMON_OTHER_LDFLAGS)` is used also _documents_ everywhere there might be an opportunity to switch to using `$(inherited)`, so I don't feel like we've discovered something and then didn't act on it.  We can just delay acting on it until later.

—

Now, having said that, I think it's okay to add `$(inherited)` in this specific case because it's a `Base.xcconfig` file.  However, there are other places where it's added for target xcconfig files, and I don't want to change those because they'll start inheriting values from that project's `Base.xcconfig` file, which usually means that `OTHER_LDFLAGS` will change.

ddkilzer replied:
BTW, here is a perfect example of an additional change that I started to add to this patch, but I decided it would be better to break out into a separate patch in case it causes an issue and needs to be rolled out separately (it still needs a review, too):  <https://github.com/WebKit/WebKit/pull/10066>

I _almost_ broke out all the `OTHER_CPLUSPLUSFLAG` deletions into their own patch as well, but I thought that change was safe enough to include here.  (I wouldn't be opposed to breaking it out, though, if someone else thought that would be best.)

emw-apple commented on Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
```diff
@@ -58,11 +58,12 @@ PRODUCT_NAME = WebCoreTestSupport;
 
 PROFILE_GENERATE_OR_USE_CFLAGS_ENABLE_LLVM_PROFILE_GENERATION = ; // Disable PGO profile generation
-OTHER_LDFLAGS[sdk=iphone*] = $(ASAN_OTHER_LDFLAGS) -lAccessibility -framework CoreText -framework Metal $(ANGLE_LDFLAGS) $(LIBWEBRTC_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+OTHER_LDFLAGS[sdk=iphone*] = $(WK_COMMON_OTHER_LDFLAGS) -lAccessibility -framework CoreText -framework Metal $(ANGLE_LDFLAGS) $(LIBWEBRTC_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```
Ditto.

ddkilzer replied:
This is a specific target xcconfig.  I don't want to change its behavior in this patch.

emw-apple commented on Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
```diff
@@ -58,11 +58,12 @@ PRODUCT_NAME = WebCoreTestSupport;
-OTHER_LDFLAGS = $(ASAN_OTHER_LDFLAGS) $(ANGLE_LDFLAGS) -framework Metal $(SOURCE_VERSION_LDFLAGS);
-OTHER_LDFLAGS[sdk=macos*] = $(ASAN_OTHER_LDFLAGS) -framework AppKit -framework Metal $(ANGLE_LDFLAGS) $(LIBWEBRTC_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) $(ANGLE_LDFLAGS) -framework Metal $(SOURCE_VERSION_LDFLAGS);
+OTHER_LDFLAGS[sdk=macos*] = $(WK_COMMON_OTHER_LDFLAGS) -framework AppKit -framework Metal $(ANGLE_LDFLAGS) $(LIBWEBRTC_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```
Ditto, plus I think you can simplify the macos line futher, because $(inherited) there will include from the non-platform-conditionalized line above:
```suggestion
OTHER_LDFLAGS = $(inherited) $(ANGLE_LDFLAGS) -framework Metal $(SOURCE_VERSION_LDFLAGS);
OTHER_LDFLAGS[sdk=macos*] = $(inherited) -framework AppKit $(LIBWEBRTC_LDFLAGS);
```

ddkilzer replied:
This is a specific target xcconfig.  I don't want to change its behavior in this patch.

I do agree this could be simplified, but this patch is already large, so I don't want to keep adding more changes to it, or it will increase the time it takes to land.  I prefer to break out changes like this into smaller patches.

emw-apple commented on Source/WebKit/Configurations/adattributiond.xcconfig:
```diff
@@ -26,14 +26,14 @@
 
 PROFILE_GENERATE_OR_USE_CFLAGS_ENABLE_LLVM_PROFILE_GENERATION = ; // Disable PGO profile generation
-OTHER_LDFLAGS = -framework WebKit $(SOURCE_VERSION_LDFLAGS) $(WK_NO_STATIC_INITIALIZERS);
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) -framework WebKit $(SOURCE_VERSION_LDFLAGS) $(WK_NO_STATIC_INITIALIZERS);
```
Ditto.

ddkilzer replied:
This is a specific target xcconfig. I don't want to change its behavior in this patch.

emw-apple commented on Source/WebKit/Configurations/webpushd.xcconfig:
```diff
@@ -26,7 +26,7 @@
 
 PROFILE_GENERATE_OR_USE_CFLAGS_ENABLE_LLVM_PROFILE_GENERATION = ; // Disable PGO profile generation
-OTHER_LDFLAGS = -framework WebKit $(SOURCE_VERSION_LDFLAGS);
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) -framework WebKit $(SOURCE_VERSION_LDFLAGS);
```
Ditto.

ddkilzer replied:
This is a specific target xcconfig. I don't want to change its behavior in this patch.

emw-apple commented on Source/WebKit/Configurations/webpushtool.xcconfig:
```diff
@@ -40,7 +40,7 @@ WK_APPSERVERSUPPORT_LDFLAGS_iphonesimulator = $(WK_APPSERVERSUPPORT_LDFLAGS$(WK_
 
 PROFILE_GENERATE_OR_USE_CFLAGS_ENABLE_LLVM_PROFILE_GENERATION = ; // Disable PGO profile generation
-OTHER_LDFLAGS = -framework Foundation -framework CoreFoundation -framework WebKit -framework JavaScriptCore $(WK_APPSERVERSUPPORT_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) -framework Foundation -framework CoreFoundation -framework WebKit -framework JavaScriptCore $(WK_APPSERVERSUPPORT_LDFLAGS) $(SOURCE_VERSION_LDFLAGS);
```
Ditto.

ddkilzer replied:
This is a specific target xcconfig. I don't want to change its behavior in this patch.

emw-apple commented on Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeApp.xcconfig:
```diff
@@ -27,7 +27,7 @@
 OTHER_LDFLAGS_FONTS = -sectcreate __DATA Ahem fonts/AHEM____.TTF -sectcreate __DATA WeightWatcher100 fonts/WebKitWeightWatcher100.ttf -sectcreate __DATA WeightWatcher200 fonts/WebKitWeightWatcher200.ttf -sectcreate __DATA WeightWatcher300 fonts/WebKitWeightWatcher300.ttf -sectcreate __DATA WeightWatcher400 fonts/WebKitWeightWatcher400.ttf -sectcreate __DATA WeightWatcher500 fonts/WebKitWeightWatc ...
 
-OTHER_LDFLAGS =  $(OTHER_LDFLAGS_$(WK_COCOA_TOUCH)) $(ASAN_OTHER_LDFLAGS);
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS) $(OTHER_LDFLAGS_$(WK_COCOA_TOUCH));
```
Ditto.

ddkilzer replied:
This is a specific target xcconfig. I don't want to change its behavior in this patch.

emw-apple commented on Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
```diff
@@ -28,7 +28,7 @@
 SKIP_INSTALL = YES;
 
-OTHER_LDFLAGS = ;
+OTHER_LDFLAGS = $(WK_COMMON_OTHER_LDFLAGS);
```
What did this do? I don't see a OTHER_LDFLAGS definition in Base or BaseTarget, so I would think it's safe to delete this altogether.

ddkilzer replied:
Good point.  I'll just delete it.

ddkilzer replied:
(BTW, I think the original intent was simply to delete all the OTHER_LDFAGS for this target.)

emw-apple commented on Tools/Scripts/webkitpy/style/checkers/basexcconfig.py:
```diff
@@ -0,0 +1,85 @@
+                if '$(inherited)' not in rhs and f'$(WK_COMMON_{name})' not in rhs:
+                    self._handle_style_error(
+                        line_number, 'basexcconfig/missing-inherited', 5,
+                        f'{name} must include \'$(inherited)\' or \'$(WK_COMMON_{name})\' (see CommonBase.xcconfig)')
```
Nice! I'm glad this worked out.

ddkilzer replied:
BTW, I think `GCC_OPTIMIZATION_LEVEL` is kind of unique among variables that may get moved into `CommonBase.xcconfig` in that it has a way to be overridden by Sanitizer builds.

Other settings could be moved into `CommonBase.xcconfig` but not have an override, and the variables appearing in a project's `Base.xcconfig` would just be a mistake, not a case where `WK_DEFAULT_` would need to be added.

So it may be best to move `GCC_OPTIMIZATION_LEVEL` into its own local variable as an overridable value, but I can't write a test non-overridden variables because we don't have an example of that yet.

Just something to keep in mind for the future.

ddkilzer replied:
> So it may be best to move GCC_OPTIMIZATION_LEVEL into its own local variable as an overridable value, but I can't write a test non-overridden variables because we don't have an example of that yet.

I went ahead and fixed this.  New patch incoming.

emw-apple commented on Tools/Scripts/webkitpy/style/checkers/basexcconfig.py:
```diff
@@ -0,0 +1,95 @@
+
+            if self.default_vars.get(name):
+                self._handle_style_error(
+                    line_number, 'basexcconfig/missing-wk_default-prefix', 5,
```
Regarding `line_number`, just want to make sure that this number is supposed to be relative to wherever `lines` occurs in the source file.

ddkilzer replied:
I believe `line_number` is supposed to match the line number for that line in the source file.

Should I not have the `start=1` parameter?

emw-apple replied:
IIRC, the checker may only give you lines that changed, so `start=1` would not always be accurate.

ddkilzer replied:
If I add this to `Configurations/CommonBase.xcconfig`:

```
CLANG_CXX_LANGUAGE_STANDARD = c++2a;
```

And then run `check-webkit-style`, the line numbers are correct.  I think it's okay to leave it now, and do a follow-up fix if it needs an update.

emw-apple commented on Tools/Scripts/webkitpy/style/checkers/basexcconfig.py:
```diff
@@ -0,0 +1,95 @@
+                    self._handle_style_error(
+                        line_number, 'basexcconfig/missing-inherited', 5,
+                        '{name} must include \'$(inherited)\' or \'$(WK_COMMON_{name})\' (see CommonBase.xcconfig)'.format(name=name))
+            elif override_vars.get(name):
+                self._handle_style_error(
+                    line_number, 'basexcconfig/overrides-common-base', 4,
+                    '{name} overrides the same variable in CommonBase.xcconfig'.format(name=name))
```
At the very least, could we suggest a workaround in this error message? My fear is that just saying "you can't override this" will encourage people to ignore the style checker.

ddkilzer replied:
I used `4` instead of `5` in `_handle_style_error()` to denote this isn't an error all the time.

I could change it to `3`, but I don't think it will be reported as an issue then.

Initially I think we should only move values into `CommonBase.xcconfig` that are intended to be the same across all projects.  If we need a way to ignore project-specific values, we can come up with that later.  (Or I could remove that check for now.)

emw-apple replied:
Keeping as-is is fine with me! We can live with this, and improve it as needed 🙂

emw-apple reviewed (approved):
From one of your comments above:

> Other settings could be moved into CommonBase.xcconfig but not have an override, and the variables appearing in a project's Base.xcconfig would just be a mistake, not a case where WK_DEFAULT_ would need to be added.

This would mean that build settings in CommonBase are "final" by default, yes? I'm not sure how I feel about that. I think we ought to expect that projects may have legitimate reasons to override any CommonBase setting, and that they should be able to do it without having to make changes to other file.

r+, but happy to continue this discussion in follow up. Thank you again!

Commit:
[Xcode] Simplify sanitizer builds
https://bugs.webkit.org/show_bug.cgi?id=244540
<rdar://98015260>

Reviewed by Elliott Williams.

This change simplifies sanitizer builds of WebKit using Xcode.
Instead of specifying `-xcconfig path/to/*san.xcconfig` on the
command-line, the ENABLE_*_SANITIZER=YES Xcode variables now
activate WebKit-specific settings.

All sanitizer settings are now in a single file named
Configurations/Sanitizers.xcconfig, which is included in every
project via Configurations/CommonBase.xcconfig.

Changes to Base.xcconfig files:
- Rename GCC_OPTIMIZATION_LEVEL variables to add WK_DEFAULT_
  prefix to support overriding in CommonBase.xcconfig.
- OTHER_CFLAGS, OTHER_CPLUSPLUS_FLAGS and OTHER_LDFAGS variables
  are defined in CommonBase.xcconfig, and per-project
  Base.xccconfig files now pick up those values through use of
  $(inherited).  When $(inherited) is not used, the
  corresponding WK_COMMON_ variable is used to include these
  values in various targets.

Other changes too numerous to list individually below:
- Rename all sanitizer variables to add WK_ prefixes.
- Rename ENABLE_ANY_SANITIZER to WK_ANY_SANITIZER_ENABLED in
  WK_NO_STATIC_INITIALIZERS variables.
- Rename WK_ASAN_DISALLOWED to WK_SANITIZER_DISALLOWED in
  yasm.xcconfig and ImageDiff.xcconfig since it disables all
  sanitizers.
- Remove ASAN_OTHER_LDFLAGS variable when the value of
  WK_SANITIZER_OTHER_LDFLAGS is inherited.
- Replace ASAN_OTHER_LDFLAGS with WK_COMMON_OTHER_LDFAGS when
  OTHER_LDFLAGS does not use $(inherited).
- Move WK_LIBCPP_ASSERTIONS_CFLAGS to be added to
  OTHER_CPLUSPLUSFLAGS in CommonBase.xcconfig.  This enables the
  macro in more projects, and removes it from OTHER_CFLAGS in
  some projects (since it only affects C++ code).

* Configurations/CommonBase.xcconfig:
- Include Sanitizers.xcconfig.
- GCC_OPTIMIZATION_LEVEL is now set (or overridden) here instead
  of in each project's Base.xcconfig.
- Initial values of OTHER_CFLAGS, OTHER_CPLUSPLUS_FLAGS and
  OTHER_LDFAGS variables are now set here.
- Move WK_COMMON_WARNING_CFLAGS and add FIXME comment.
* Configurations/SDKVariant.xcconfig:
- Add WK_ADD_ macros used by Sanitizers.xcconfig.
* Configurations/Sanitizers.xcconfig: Add.
- Combines all of the Tools/sanitizer xcconfig files into one.
- Add -fsanitize=enum,return for UBSan since some versions of
  Xcode disable these checks by default.
- Add a FIXME for disabling sanitizers on host-side tools.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
- Switch OTHER_LDFLAGS to use $(inherited).
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/ThirdParty/libwebrtc/Configurations/yasm.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/libavif.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/dav1d.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/Configurations/webpushd.xcconfig:
* Source/WebKit/Configurations/webpushtool.xcconfig:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/DumpRenderTree/mac/Configurations/BaseTarget.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeApp.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
- Delete OTHER_LDFLAGS since it was not overriding anything from
  its Base.xcconfig.
* Tools/ImageDiff/cg/Configurations/ImageDiff.xcconfig:
- Remove redundant OTHER_CPLUSPLUSFLAGS variable.
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/WebGPUPlayground/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

* Tools/Scripts/webkitdirs.pm:
(XcodeOptions):
- Simplify options for enabling sanitizers or overriding the
  compiler optimization level.
* Tools/Scripts/webkitpy/style/checker.py:
(_all_categories):
- Add BaseXcconfigChecker categories.
(FileType):
- Replace unused SDKVARIANT with BASE_XCCONFIG.
(CheckerDispatcher._file_type):
- Return FileType for Base.xcconfig files.
(CheckerDispatcher._create_checker):
- Set checker for Base.xcconfig files.
* Tools/Scripts/webkitpy/style/checker_unittest.py:
(CheckerDispatcherDispatchTest.assert_checker_basexcconfig):
(CheckerDispatcherDispatchTest.test_basexcconfig_paths):
- Add test for BaseXcconfigChecker.
* Tools/Scripts/webkitpy/style/checkers/basexcconfig.py: Add.
(BaseXcconfigChecker):
(BaseXcconfigChecker.__init__):
(BaseXcconfigChecker.read_common_base_xcconfig_variables):
(BaseXcconfigChecker.check):
- Implement BaseXcconfigChecker.
- The checker is dynamic in that it reads CommonBase.xcconfig to
  determine which variables require checking in Base.xcconfig
  project files.
* Tools/Scripts/webkitpy/style/checkers/basexcconfig_unittest.py: Add.
(BaseXcconfigCheckerTest):
(BaseXcconfigCheckerTest.assert_no_error):
(BaseXcconfigCheckerTest.assert_no_error.handle_style_error):
(BaseXcconfigCheckerTest.assert_error):
(BaseXcconfigCheckerTest.assert_error.handle_style_error):
(BaseXcconfigCheckerTest.mock_handle_style_error):
(BaseXcconfigCheckerTest.test_init):
(BaseXcconfigCheckerTest.test_gcc_optimization_level):
(BaseXcconfigCheckerTest.test_wk_default_gcc_optimization_level):
(BaseXcconfigCheckerTest.test_other_cflags):
(BaseXcconfigCheckerTest.test_other_cplusplusflags):
(BaseXcconfigCheckerTest.test_other_ldflags):
- Add unit tests for BaseXcconfigChecker.

* Tools/sanitizer/asan+ubsan.xcconfig: Remove.
* Tools/sanitizer/asan.xcconfig: Remove.
* Tools/sanitizer/sanitizer.xcconfig: Remove.
* Tools/sanitizer/tsan.xcconfig: Remove.
* Tools/sanitizer/ubsan.xcconfig: Remove.

Canonical link: https://commits.webkit.org/260654@main
